### PR TITLE
RFR - Use Authed APIs for internal trigger registration

### PR DESCRIPTION
**Pull request by lakshmi-kannan**
With this, also refactored code into a class. Config is not read during import time. This should save us from import dependencies. 

Seems to work in dev env. I am going to see if I can add some tests for the class. 

```
~/st2 git:master ❯❯❯ mongo                                                                                             ✭ ✱ ◼
MongoDB shell version: 2.6.3
connecting to: test
> use st2
switched to db st2
> show db.trigger_d_b.find()
2015-04-28T05:39:10.278+0000 don't know how to show [db.trigger_d_b.find()] at src/mongo/shell/utils.js:729
> show db.trigger_d_b.find()
2015-04-28T05:39:13.094+0000 don't know how to show [db.trigger_d_b.find()] at src/mongo/shell/utils.js:729
> db.trigger_d_b.find()
{ "_id" : ObjectId("553f1cd56d0aff6cae840d2b"), "name" : "polling_event", "pack" : "examples", "type" : "examples.polling_event", "parameters" : {  } }
{ "_id" : ObjectId("553f1cd56d0aff6cae840d2e"), "name" : "event", "pack" : "examples", "type" : "examples.event", "parameters" : {  } }
{ "_id" : ObjectId("553f1cd66d0aff6c889ff253"), "name" : "st2.generic.actiontrigger", "pack" : "core", "type" : "core.st2.generic.actiontrigger", "parameters" : {  } }
{ "_id" : ObjectId("553f1cd66d0aff6c889ff255"), "name" : "st2.generic.notifytrigger", "pack" : "core", "type" : "core.st2.generic.notifytrigger", "parameters" : {  } }
{ "_id" : ObjectId("553f1cd66d0aff6c889ff254"), "name" : "st2.sensor.process_spawn", "pack" : "core", "type" : "core.st2.sensor.process_spawn", "parameters" : {  } }
{ "_id" : ObjectId("553f1cd66d0aff6c889ff256"), "name" : "st2.sensor.process_exit", "pack" : "core", "type" : "core.st2.sensor.process_exit", "parameters" : {  } }
{ "_id" : ObjectId("553f1cd66d0aff6cae840d6c"), "name" : "dff2ae4c-3f8b-44ad-9b65-8c8afe541637", "pack" : "core", "type" : "core.st2.IntervalTimer", "parameters" : { "unit" : "seconds", "delta" : 5 } }
{ "_id" : ObjectId("553f1cd66d0aff6cae840d6e"), "name" : "1cdc5709-281c-44a2-b582-44586744dc1a", "pack" : "core", "type" : "core.st2.webhook", "parameters" : { "url" : "sample" } }
>
```

### CODE REVIEW

File: st2common/st2common/triggers.py [hunk 1]
```diff
@@ -32,75 +33,83 @@
...
 
+class InternalTriggerTypesRegistrar(object):
+
+    def __init__(self):
+        self._action_sensor_enabled = cfg.CONF.action_sensor.enable
+        self._trigger_type_endpoint = cfg.CONF.action_sensor.triggers_base_url
+        self._retry_wait = cfg.CONF.action_sensor.retry_wait
+        self._timeout = cfg.CONF.action_sensor.request_timeout
+        self._max_attempts = cfg.CONF.action_sensor.max_attempts
+        self._temporary_token = create_token('system.internal_trigger_registrar',
```
Comment (Kami) [inline st2common/st2common/triggers.py]: We should use a smaller TTL. Something like 5 min should be sufficient here.

Comment (lakshmi-kannan) [inline st2common/st2common/triggers.py]: Yeah. Set it to one hour in my recent commit. We anyways delete the token at the end. 5 mins might be too small IMO. We'll have catch unauthorized code and renegotiate a token. Seems too much code :P.

File: st2common/st2common/triggers.py [hunk 2]
```diff
@@ -32,75 +33,85 @@
...
-HTTP_POST_HEADER = {'content-type': 'application/json'}
-RETRY_WAIT = cfg.CONF.action_sensor.retry_wait
-TIMEOUT = cfg.CONF.action_sensor.request_timeout
-MAX_ATTEMPTS = cfg.CONF.action_sensor.max_attempts
 
+class InternalTriggerTypesRegistrar(object):
+
+    def __init__(self):
+        self._action_sensor_enabled = cfg.CONF.action_sensor.enable
+        self._trigger_type_endpoint = cfg.CONF.action_sensor.triggers_base_url
```
Comment (Kami) [inline st2common/st2common/triggers.py]: Eventually we should probably switch to `auth.api_url` setting so we don't need to have 10 different settings which basically refer to the same thing.

File: st2actions/st2actions/cmd/actionrunner.py [hunk 3]
```diff
@@ -46,9 +46,6 @@ def _setup():
     register_common_signal_handlers()
 
     # 4. Register internal triggers
-    # Note: We need to do import here because of a messed up configuration
```
Comment (manasdk) [inline st2actions/st2actions/cmd/actionrunner.py]: importing in the right order makes the config issue go away?

Crap. This gives me c/c++ import order nightmares all over again!

File: st2common/st2common/triggers.py [hunk 4]
```diff
@@ -32,75 +33,85 @@
...
+        self._retry_wait = cfg.CONF.action_sensor.retry_wait
+        self._timeout = cfg.CONF.action_sensor.request_timeout
+        self._max_attempts = cfg.CONF.action_sensor.max_attempts
+        self._auth_creds = create_token('system.internal_trigger_registrar',
+                                        ttl=(1 * 60 * 60))
```
Comment (Kami) [inline st2common/st2common/triggers.py]: `ttl` is in seconds, right? If so, I assume we probably want 60 seconds (1 minute) and not one hour.

Or was this intentional?

File: st2common/st2common/triggers.py [hunk 5]
```diff
@@ -32,75 +33,85 @@
...
+        self._http_post_headers = {'content-type': 'application/json',
+                                   'X-Auth-Token': self._auth_creds.token}
+        self._get_headers = {'X-Auth-Token': self._auth_creds.token}
+
+    def register_internal_trigger_types(self):
+        LOG.debug('Registering internal trigger types...')
+
+        for resource_name, trigger_definitions in INTERNAL_TRIGGER_TYPES.items():
+            for trigger_definition in trigger_definitions:
+                LOG.debug('Registering internal trigger: %s', trigger_definition['name'])
```
Comment (Kami) [inline st2common/st2common/triggers.py]: That's fine for now, but honestly, we should probably just get rid of this setting and just always publish this trigger.

File: st2common/st2common/triggers.py [hunk 6]
```diff
@@ -32,75 +33,85 @@
...
 
+class InternalTriggerTypesRegistrar(object):
+
+    def __init__(self):
+        self._action_sensor_enabled = cfg.CONF.action_sensor.enable
+        self._trigger_type_endpoint = cfg.CONF.action_sensor.triggers_base_url
+        self._retry_wait = cfg.CONF.action_sensor.retry_wait
+        self._timeout = cfg.CONF.action_sensor.request_timeout
+        self._max_attempts = cfg.CONF.action_sensor.max_attempts
+        self._auth_creds = create_token('system.internal_trigger_registrar',
```
Comment (manasdk) [inline st2common/st2common/triggers.py]: do we tombstone tokens?

File: st2common/st2common/triggers.py [hunk 7]
```diff
@@ -32,75 +33,85 @@
...
+                                        ttl=(1 * 60 * 60))
+        self._http_post_headers = {'content-type': 'application/json',
+                                   'X-Auth-Token': self._auth_creds.token}
+        self._get_headers = {'X-Auth-Token': self._auth_creds.token}
```
Comment (manasdk) [inline st2common/st2common/triggers.py]: _http_get_headers?

File: st2actions/st2actions/cmd/actionrunner.py (hunk 3 above)
Comment (Kami) [inline st2actions/st2actions/cmd/actionrunner.py]: Yeah, our bootstraping phase is pretty messed up right now. Known issue for quite a while now.

File: st2common/st2common/triggers.py [hunk 8]
```diff
@@ -32,75 +33,85 @@
...
+
+        for resource_name, trigger_definitions in INTERNAL_TRIGGER_TYPES.items():
+            for trigger_definition in trigger_definitions:
+                LOG.debug('Registering internal trigger: %s', trigger_definition['name'])
+                if (trigger_definition['name'] == ACTION_SENSOR_TRIGGER['name'] and
+                        not self._action_sensor_enabled):
+                    continue
+                self._register_trigger_type(trigger_definition=trigger_definition, attempt_no=0)
+
+        delete_token(self._auth_creds.token)
```
Comment (manasdk) [inline st2common/st2common/triggers.py]: If you delete_token here then create in this method also.

Comment (manasdk) [inline st2common/st2common/triggers.py]: crap you have self._get_headers etc littered around. Ugh. Fine. Let it be.

File: st2common/tests/unit/test_internal_trigger_types_registrar.py [hunk 9]
```diff
@@ -0,0 +1,74 @@
...
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an 'AS IS' BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import datetime
+try:
+    import simplejson as json
```
Comment (manasdk) [inline st2common/tests/unit/test_internal_trigger_types_registrar.py]: simplejson isnt a dep so why do we always do this?

File: st2common/st2common/triggers.py [hunk 10]
```diff
@@ -32,75 +33,85 @@
...
+        self._auth_creds = create_token('system.internal_trigger_registrar',
+                                        ttl=(1 * 60 * 60))
+        self._http_post_headers = {'content-type': 'application/json',
+                                   'X-Auth-Token': self._auth_creds.token}
+        self._get_headers = {'X-Auth-Token': self._auth_creds.token}
+
+    def register_internal_trigger_types(self):
+        LOG.debug('Registering internal trigger types...')
+
+        for resource_name, trigger_definitions in INTERNAL_TRIGGER_TYPES.items():
```
Comment (manasdk) [inline st2common/st2common/triggers.py]: six

Comment (Kami) [inline st2common/st2common/triggers.py]: six is only required for `iteritems`, `items` is available in 2.x and 3.x, but it's true that they behave differently :P

File: st2common/st2common/triggers.py (hunk 4 above)
Comment (lakshmi-kannan) [inline st2common/st2common/triggers.py]: This was intentional.

Comment (Kami) [inline st2common/st2common/triggers.py]: K.

File: st2common/st2common/triggers.py (hunk 5 above)
Comment (lakshmi-kannan) [inline st2common/st2common/triggers.py]: I am +1 to this. I just wanted to not change too many things in this PR.

File: st2common/st2common/triggers.py (hunk 8 above)
Comment (lakshmi-kannan) [inline st2common/st2common/triggers.py]: Ideally delete_token happens in the destructor for this registrar. C++ times :P.

File: st2common/tests/unit/test_internal_trigger_types_registrar.py (hunk 9 above)
Comment (lakshmi-kannan) [inline st2common/tests/unit/test_internal_trigger_types_registrar.py]: Dude, I am linking this to you one more time. Next time you ask me about this, our relationship won't be the same :P.
http://stackoverflow.com/questions/712791/what-are-the-differences-between-json-and-simplejson-python-modules

File: st2common/st2common/triggers.py (hunk 6 above)
Comment (lakshmi-kannan) [inline st2common/st2common/triggers.py]: Dude, this deletes the db object. No tombstoning :/.

Comment (manasdk) [inline st2common/st2common/triggers.py]: Yeah I figured that part later

File: st2common/st2common/triggers.py (hunk 8 above)
Comment (manasdk) [inline st2common/st2common/triggers.py]: destructor. lol!

Comment (Kami) [inline st2common/st2common/triggers.py]: __del__, well, it doesn't really work as a destructor in Python and it's a bad idea to use it :P

**Pull request merged**